### DISCUSSION TITLE
Allow upserting relations

### DISCUSSION THREAD

**Discussion by Kalos-S** [mikro-orm/mikro-orm#4708] [General]
**Is your feature request related to a problem? Please describe.**
When using upsert to create or update entities a ForeignKeyConstraintViolationException is thrown when attempting to upsert a entity with a required relation. This is because the em.create call creates both the entity and its relation but the em.upsert call expects all relations to already be present in the database.

**Describe the solution you'd like**
Handle relations in a similar way to em.flush

**Describe alternatives you've considered**
Creating custom upsert logic that loads the entity from the database and performs different actions based on whether it exists or not is possible, results in lots of duplicate code though.

**Comment by B4nan** [comment]
Please provide some code, not sure what exactly are you doing, but you dont need `em.create` for upserting at all. 

> Handle relations in a similar way to em.flush

Not sure how you expect this to work?

**Comment by Kalos-S** [comment]
Well, code like this:
```typescript
const entity = em.create(Entity, {
  foo: ”bar”,
  relation: {
    bar: ”baz“,
  },
});

await em.upsert(entity);
```

Tries to insert the entity with the values { foo: ”bar”, relation: ”Primary Key of relation, created in em.create()” }. If the relation is not found in the database, this results in a ForeignKeyConstraintViolationException. Similarly, if em.create() is not used the entity is incomplete since no constructor is used.

**Comment by Kalos-S** [comment]
I guess another way of describing this is if an unmanaged relation is used in an upsert call, first upsert that relation and then upsert the entity itself. The previous example could probably be fixed (Not able to test yet) by calling

```typescript
await em.upsert(entity.relation);
await em.upsert(entity);
```

**Reply by B4nan** [reply]
I see, yeah I think that makes sense, at least for m:1 and 1:1 owning sides, maybe for inverse sides too but in reversed order, will need to test that but it feels like the same algo as `UoW.commit()` is using might be needed, and I would rather not duplicate such a complex thing. Maybe `em.flush({ upsert: true })` could do the job, over time quite a lot of people were asking for something like it.

Although most of the newly added upsert options won't work for the relations, e.g. `onConflictFields`, as that would point to the root entity properties. Maybe we could allow setting those on the metadata level (e.g. via `@Entity` decorator).

Overall this could end up as a pretty big feature, so will keep that for v6, need to focus on that now.

**Comment by novikov1337danil** [comment]
Have there been any updates for this feature?

**Reply by novikov1337danil** [reply]
I received a [response](https://discord.com/channels/1214904142443839538/1394620504794726441) from @B4nan that this has not been done now, but it may be implemented in some parts of v7
